### DISCUSSION TITLE
Using artifactory with Renovate does not provide json metadata

### DISCUSSION THREAD

**Discussion by bavedarnow** [renovatebot/renovate#7971] [Request Help] [answered]
I've managed to get renovate working with our internal artifactory repo, only to realize that the change log URLs aren't coming through in the pull requests. This is because artifactory [does not support](https://www.jfrog.com/jira/browse/RTFACT-16101) the `/json` endpoint and therefore all of the metadata is unavailable for renovate to parse. Since artifactory is just a remote cache of PyPi, is there a way to grab the json data from PyPi? Potentially by switching out the lookupUrl defined on [this line](https://github.com/renovatebot/renovate/blob/f900b70147a8a32ead09d4708b101f2864e455f0/lib/datasource/pypi/index.ts#L66)?

Alternatively, is there a way to define a pip index for our private packages only? I'm using the `PIP_INDEX_URL` environment variable with renovate to define which index to use.

Thanks in advance!

**Comment by rarkins** [comment]
> Alternatively, is there a way to define a pip index for our private packages only? I'm using the `PIP_INDEX_URL` environment variable with renovate to define which index to use.

Yes, you could remove that environment variable and instead just direct lookups for particular packages using `packageRules` with `packageNames` and `registryUrls`. i.e. if packageNames includes X then use registryUrls Y

**Reply by bavedarnow** [reply]
I'm trying this out and realizing this wont work because we're using environment variables to feed HTTP basic auth username/password to the `PIP_INDEX_URL` within jenkins. As far as I know, there isn't a way to use environment variables in renovate configs, right?

```groovy
pipeline {
  agent any
  environment {
    PATH="/home/user/.yarn/bin:/usr/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:$PATH"
    RENOVATE_CONFIG_FILE="renovate-config.js"
    RENOVATE_TOKEN=credentials("XXX")
    GITHUB_COM_TOKEN=credentials("XXX")
    ARTIFACTORY = credentials("artifactory")
    ARTIFACTORY_REPO="https://$ARTIFACTORY_USR:$ARTIFACTORY_PSW@the_artifactory_url/simple"
  }
  triggers {
    cron('@hourly')
  }
  stages {
    stage('Renovate') {
      agent {
        dockerfile {
          label 'docker'
        }
      }

      steps {
        sh "PIP_INDEX_URL=$env.ARTIFACTORY_REPO renovate"
        sh "cp /tmp/renovate.log ${WORKSPACE}"
      }

      post {
        always {
          archiveArtifacts 'renovate.log'
        }
      }
    }
  }
}
```

**Reply by viceice** [reply]
you can use them in your `renovate-config.js` like `process.env.ARTIFACTORY` and assign them to a host rule

**Comment by bavedarnow** [comment, marked as answer]
After lots of trial and error, I was able to get this working! Thanks for the help.

`Jenkinsfile`:
```groovy
pipeline {
  agent any
  environment {
    PATH="/home/user/.yarn/bin:/usr/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:$PATH"
    RENOVATE_CONFIG_FILE="renovate-config.js"
    RENOVATE_TOKEN=credentials("XXX")
    GITHUB_COM_TOKEN=credentials("XXX")
    ARTIFACTORY = credentials("artifactory")
  }
  triggers {
    cron('@hourly')
  }
  stages {
    stage('Renovate') {
      agent {
        dockerfile { }
      }

      steps {
        sh "renovate"
      }
    }
  }
}
```
`renovate-config.js`:
```js
module.exports = {
  packageRules: [
    {
      packageNames: ['your', 'artifactory', 'packages'],
      registryUrls: ['https://artifactory.yourcompany.com/repo-name/simple/']
    }
  ],
  hostRules: [
    {
      baseUrl: 'https://artifactory.yourcompany.com/',
      username: process.env.ARTIFACTORY_USR,
      password: process.env.ARTIFACTORY_PSW
    }
  ]
};
```
